File: apps/desktop/src-tauri/src/commands.rs
pub const APPLICATION_PLUGIN_NAME: &str = "application";

pub const AUTH_PLUGIN_NAME: &str = "auth";
pub const EVENTS_PLUGIN_NAME: &str = "events";
pub const INSTANCE_PLUGIN_NAME: &str = "instance";
pub const MINECRAFT_PLUGIN_NAME: &str = "minecraft";
pub const PLUGIN_PLUGIN_NAME: &str = "plugin";
pub const PROCESS_PLUGIN_NAME: &str = "process";
pub const SETTINGS_PLUGIN_NAME: &str = "settings";
pub const UPDATE_PLUGIN_NAME: &str = "update";

macro_rules! application_commands {
    ($($tokens:tt)*) => {
        $($tokens)* [
            initialize_state,
            initialize_plugins,
        ]
    };
}

macro_rules! auth_commands {
    ($($tokens:tt)*) => {
        $($tokens)* [
            list_accounts,
            create_offline_account,
            change_account,
            logout,
        ]
    };
}

macro_rules! events_commands {
    ($($tokens:tt)*) => {
        $($tokens)* [
            list_progress_bars,
        ]
    };
}

macro_rules! instance_commands {
    ($($tokens:tt)*) => {
        $($tokens)* [
            create,
            import,
            list_importers,
            list,
            get,
            get_dir,
            install,
            update,
            edit,
            remove,
            launch,
            stop,
            import_contents,
            list_content,
            install_content,
            enable_contents,
            disable_contents,
            remove_contents,
            list_content_providers,
            search_content,
            check_compatibility,
            get_content,
            list_content_version,
        ]
    };
}

macro_rules! minecraft_commands {
    ($($tokens:tt)*) => {
        $($tokens)* [
            get_minecraft_version_manifest,
            get_loader_version_manifest,
        ]
    };
}

macro_rules! plugin_commands {
    ($($tokens:tt)*) => {
        $($tokens)* [
            import,
            sync,
            list,
            get,
            remove,
            enable,
            disable,
            call,
            get_settings,
            edit_settings,
            open_plugins_folder,
            get_api_version,
        ]
    };
}

macro_rules! process_commands {
    ($($tokens:tt)*) => {
        $($tokens)* [
            list,
            get_by_instance_id,
        ]
    };
}

macro_rules! settings_commands {
    ($($tokens:tt)*) => {
        $($tokens)* [
            get,
            edit,
            get_max_ram,
            get_default_instance_settings,
            edit_default_instance_settings,
            get_app_settings,
            edit_app_settings,
        ]
    };
}

macro_rules! update_commands {
    ($($tokens:tt)*) => {
        $($tokens)* [
            check_for_updates,
            install_update,
        ]
    };
}

pub(crate) use application_commands;

pub(crate) use auth_commands;
pub(crate) use events_commands;
pub(crate) use instance_commands;
pub(crate) use minecraft_commands;
pub(crate) use plugin_commands;
pub(crate) use process_commands;
pub(crate) use settings_commands;
pub(crate) use update_commands;

File: apps/ui/src/shared/api/commands.ts
import { commands as applicationRawCommands } from './bindings/application';
import { commands as authRawCommands } from './bindings/auth';
import {
  commands as eventsRawCommands,
  events as eventsRawEvents,
} from './bindings/events';
import {
  commands as instanceRawCommands,
  events as instanceRawEvents,
} from './bindings/instance';
import { commands as javaRawCommands } from './bindings/java';
import { commands as minecraftRawCommands } from './bindings/minecraft';
import {
  commands as pluginsRawCommands,
  events as pluginsRawEvents,
} from './bindings/plugin';
import {
  commands as processRawCommands,
  events as processRawEvents,
} from './bindings/process';
import { commands as settingsRawCommands } from './bindings/settings';
import { commands as updateRawCommands } from './bindings/update';
import { withIdempotency } from './tauri';

export const authCommands = {
  listAccounts: authRawCommands.listAccounts,
  createOfflineAccount: withIdempotency(authRawCommands.createOfflineAccount),
  changeAccount: withIdempotency(authRawCommands.changeAccount),
  logout: withIdempotency(authRawCommands.logout),
};

export const updateCommands = {
  checkForUpdates: updateRawCommands.checkForUpdates,
  installUpdate: withIdempotency(updateRawCommands.installUpdate),
};

export const pluginsCommands = {
  import: withIdempotency(pluginsRawCommands.import),
  sync: withIdempotency(pluginsRawCommands.sync),
  list: pluginsRawCommands.list,
  get: pluginsRawCommands.get,
  remove: withIdempotency(pluginsRawCommands.remove),
  enable: withIdempotency(pluginsRawCommands.enable),
  disable: withIdempotency(pluginsRawCommands.disable),
  call: withIdempotency(pluginsRawCommands.call),
  getSettings: pluginsRawCommands.getSettings,
  editSettings: withIdempotency(pluginsRawCommands.editSettings),
  openPluginsFolder: pluginsRawCommands.openPluginsFolder,
  getApiVersion: pluginsRawCommands.getApiVersion,
};

export const settingsCommands = {
  get: settingsRawCommands.get,
  edit: withIdempotency(settingsRawCommands.edit),
  getMaxRam: settingsRawCommands.getMaxRam,
  getDefaultInstanceSettings: settingsRawCommands.getDefaultInstanceSettings,
  editDefaultInstanceSettings: withIdempotency(
    settingsRawCommands.editDefaultInstanceSettings,
  ),
  getAppSettings: settingsRawCommands.getAppSettings,
  editAppSettings: withIdempotency(settingsRawCommands.editAppSettings),
};

export const instanceCommands = {
  create: withIdempotency(instanceRawCommands.create),
  import: withIdempotency(instanceRawCommands.import),
  listImporters: instanceRawCommands.listImporters,
  list: instanceRawCommands.list,
  get: instanceRawCommands.get,
  getDir: instanceRawCommands.getDir,
  install: withIdempotency(instanceRawCommands.install),
  update: withIdempotency(instanceRawCommands.update),
  edit: withIdempotency(instanceRawCommands.edit),
  remove: withIdempotency(instanceRawCommands.remove),
  launch: withIdempotency(instanceRawCommands.launch),
  stop: withIdempotency(instanceRawCommands.stop),
  importContents: withIdempotency(instanceRawCommands.importContents),
  listContent: instanceRawCommands.listContent,
  installContent: withIdempotency(instanceRawCommands.installContent),
  enableContents: withIdempotency(instanceRawCommands.enableContents),
  disableContents: withIdempotency(instanceRawCommands.disableContents),
  removeContents: withIdempotency(instanceRawCommands.removeContents),
  listContentProviders: instanceRawCommands.listContentProviders,
  searchContent: instanceRawCommands.searchContent,
  checkCompatibility: instanceRawCommands.checkCompatibility,
  getContent: instanceRawCommands.getContent,
  listContentVersion: instanceRawCommands.listContentVersion,
  editIcon: withIdempotency(instanceRawCommands.editIcon),
};

export const javaCommands = {
  list: javaRawCommands.list,
  edit: withIdempotency(javaRawCommands.edit),
  remove: withIdempotency(javaRawCommands.remove),
  install: withIdempotency(javaRawCommands.install),
  testJre: javaRawCommands.testJre,
  discover: javaRawCommands.discover,
  getActiveInstallations: javaRawCommands.getActiveInstallations,
};

export const minecraftCommands = {
  getMinecraftVersionManifest: minecraftRawCommands.getMinecraftVersionManifest,
  getLoaderVersionManifest: minecraftRawCommands.getLoaderVersionManifest,
};

export const processCommands = {
  list: processRawCommands.list,
  getByInstanceId: processRawCommands.getByInstanceId,
};

export const eventsCommands = {
  listProgressBars: eventsRawCommands.listProgressBars,
};

export const applicationCommands = {
  waitForInitialization: applicationRawCommands.waitForInitialization,
  initializePlugins: applicationRawCommands.initializePlugins,
  recreateWindow: applicationRawCommands.recreateWindow,
};

// Re-export events from bindings
export const pluginsEvents = pluginsRawEvents;
export const instanceEvents = instanceRawEvents;
export const processEvents = processRawEvents;
export const eventsEvents = eventsRawEvents;
